File: JavaScript/6js 1/deepCount.js
/*
|--------------------------------------------------------------------------
| Array Deep Count
|--------------------------------------------------------------------------
|
*/

function deepCount(a) {
   let count = 0;
   function countHelper(arr) {
      count += arr.length;
      for (let element of arr) {
         if (Array.isArray(element)) countHelper(element);
      }
   }
   countHelper(a);
   return count;
}


function deepCount(a) {
   return a.reduce((s, e) => s + (Array.isArray(e) ? deepCount(e) : 0), a.length);
}


function deepCount(a) {
   let count = a.length;
   for (let i = 0; i < a.length; i++) if (Array.isArray(a[i])) count += deepCount(a[i]);
   return count;
}


console.log(deepCount([]), 0, "Expected 0");
console.log(deepCount([1, 2, 3]), 3, "Expected 3");
console.log(deepCount(["x", "y", ["z"]]), 4, "Expected 4");
console.log(deepCount([1, 2, [3, 4, [5]]]), 7, "Expected 7");
console.log(deepCount([[[[[[[[[]]]]]]]]]), 8, "Expected 8");

File: Python/6py 1/deep_count.py
"""
|--------------------------------------------------------------------------
| Array Deep Count
|--------------------------------------------------------------------------
|
"""

def deep_count(a):
   count = 0
   def count_helper(arr):
      nonlocal count
      count += len(arr)
      for element in arr:
         if isinstance(element, list): count_helper(element)
   count_helper(a)
   return count


deep_count=d=lambda a:sum(type(e)is not list or d(e)+1for e in a)


def deep_count(a):
   return sum(1 + (deep_count(x) if isinstance(x, list) else 0) for x in a)


print(deep_count([]), 0)
print(deep_count([1, 2, 3]), 3)
print(deep_count(["x", "y", ["z"]]), 4)
print(deep_count([1, 2, [3, 4, [5]]]), 7)
print(deep_count([[[[[[[[[]]]]]]]]]), 8)